File: LinkList/19_Remove_Nth_Node_From_End_of_List.cpp
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        vector<ListNode*>m(1, NULL);/*space O(n)*/
        ListNode *move = head;
        int count = 0;
        while(move!=NULL){ /*Time O(n)*/
            count++;
            m.push_back(move);
            move = move->next;
        }
        m.push_back(NULL);
        int index = count - n;
        if(index == 0) return head->next;
        m[index]->next = m[index+2];
        return head;
    }
};

/*Another approach without using extra space*/

File: Practice_Algorithm/19. Remove Nth Node From End of List.java
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode s = head, f = head, pre = head;
        for(int i = 0; i < n; i++){
            f = f.next;
        }
        while(f!=null){
            pre = s;
            s = s.next;
            f = f.next;
        }
        pre.next = s.next;
        return pre==s?head.next:head;
    }
}
